### PR TITLE
Deploying Sam App calls sam build, updates statusbar

### PR DESCRIPTION
**Pull request by awschristou**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Infrastructure (change which improves the lifecycle management (CI/CD, build, package, deploy, lint, etc) of the application, but does not change functionality.)
- [ ] Technical debt (change which improves the maintainability of the codebase, but does not change functionality)
- [ ] Testing (change which modifies or adds test coverage, but does not change functionality)
- [ ] Documentation (change which modifies or adds documentation, but does not change functionality)

## Description

* Added a status bar entry that persists for the duration of a SAM App deploy operation:
image
* Deploying Sam apps now calls `sam build` prior to package and deploy, so that dependencies are pulled in.

## Related Issue(s)
#274 

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: package.nls.json [hunk 1]
```diff
@@ -120,6 +120,7 @@
     "AWS.samcli.deploy.stackName.error.invalidCharacters": "A stack name may contain only alphanumeric characters (case sensitive) and hyphens",
     "AWS.samcli.deploy.stackName.error.firstCharacter": "A stack name must begin with an alphabetic character",
     "AWS.samcli.deploy.stackName.error.length": "A stack name must not be longer than 128 characters",
+    "AWS.samcli.deploy.statusbar.message": "$(cloud-upload) Deploying SAM Application to {0}...",
```
Comment (mpiroc) [inline package.nls.json]: What does the syntax `$(cloud-upload)` do?

File: src/lambda/commands/deploySamApplication.ts [hunk 2]
```diff
@@ -30,40 +31,71 @@ export async function deploySamApplication({
...
-        const packageInvocation = new SamCliPackageInvocation(template.fsPath, outputTemplatePath, s3Bucket, invoker)
-        restParams.outputChannel.show(true)
-        // TODO: Add nls support
-        restParams.outputChannel.appendLine(`Packaging SAM Application to S3 Bucket: ${s3Bucket}`)
-        await packageInvocation.execute()
+        const tempFolder = await mkdtemp('samDeploy')
+        const buildDestination = path.join(tempFolder, 'build')
+        const buildTemplatePath = path.join(buildDestination, 'template.yaml')
+        const outputTemplatePath = path.join(tempFolder, 'template.yaml')
+        let stage = 'starting up'
```
Comment (mpiroc) [inline src/lambda/commands/deploySamApplication.ts]: Nit: Suggest to make `stage` type `'starting up' | 'building' | 'packaging'` to protect against typos.

File: src/lambda/commands/deploySamApplication.ts [hunk 3]
```diff
@@ -30,40 +31,71 @@ export async function deploySamApplication({
...
+        try {
+            restParams.outputChannel.show(true)
+            stage = 'building'
+            const build = new SamCliBuildInvocation(
+                buildDestination,
+                undefined,
+                template.fsPath,
+                invoker
+            )
+            // TODO: Add nls support
```
Comment (mpiroc) [inline src/lambda/commands/deploySamApplication.ts]: We should probably have a Release Readiness task for ensuring that all of our strings are localizable.

File: src/lambda/commands/deploySamApplication.ts [hunk 4]
```diff
@@ -30,40 +31,71 @@ export async function deploySamApplication({
...
+        const buildDestination = path.join(tempFolder, 'build')
+        const buildTemplatePath = path.join(buildDestination, 'template.yaml')
+        const outputTemplatePath = path.join(tempFolder, 'template.yaml')
+        let stage = 'starting up'
+        try {
+            restParams.outputChannel.show(true)
+            stage = 'building'
+            const build = new SamCliBuildInvocation(
+                buildDestination,
+                undefined,
```
Comment (mpiroc) [inline src/lambda/commands/deploySamApplication.ts]: What's the purpose of omitting the build dir? Is this sensitive to the CWD?

File: src/lambda/commands/deploySamApplication.ts [hunk 5]
```diff
@@ -30,40 +31,71 @@ export async function deploySamApplication({
     }
 
     const { template, s3Bucket, stackName } = args
+    const deployApplicationPromise = new Promise<void>(async (resolve, reject) => {
```
Comment (mpiroc) [inline src/lambda/commands/deploySamApplication.ts]: I don't think we need to manually create a promise here, just use an `async (): Promise<void>` method instead.

File: package.nls.json (hunk 1 above)
Comment (awschristou) [reply package.nls.json]: It renders the icon you can see in the PR description screenshot
info: https://code.visualstudio.com/api/references/icons-in-labels

File: src/lambda/commands/deploySamApplication.ts (hunk 4 above)
Comment (awschristou) [reply src/lambda/commands/deploySamApplication.ts]: We omit the base directory.

The template can have multiple resources, and you cannot specify a base directory per resource. At this point you're relying on the template (and all resource CodeUris) to be correctly configured.

From the [docs](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-cli-command-reference-sam-build.html), "By default, relative paths are resolved with respect to the template's location."

File: src/lambda/commands/deploySamApplication.ts (hunk 5 above)
Comment (awschristou) [reply src/lambda/commands/deploySamApplication.ts]: I'd prefer not to put that directly into the `setStatusBarMessage` call, it feels harder to read.

Comment (mpiroc) [reply src/lambda/commands/deploySamApplication.ts]: You don't need to, just replace:

```typescript
const deployApplicationPromise = new Promise<void>(async (resolve, reject) => {
    // ....
})
```

with

```typescript
const deployApplicationPromise = async () => {
    // ...
}()
```

Up to you though.

**Pull request merged**